File: 0067-add-binary/0067-add-binary.java
class Solution {
    public String addBinary(String a, String b) {

        //Option 3) using Bit manipulation
        /*
        int x= Integer.parseInt(a, 2);
        int y = Integer.parseInt(b, 2);

        while (y != 0) {
            int temp = (x & y) << 1;
            x= x ^ y;
            y = temp;
        }

       return Integer.toBinaryString(x);  */

        //Option 2
        //return Integer.toBinaryString(Integer.parseInt(a, 2) + Integer.parseInt(b, 2));

        //option1()
         StringBuilder sb = new StringBuilder();
        int i = a.length() - 1;
        int j = b.length() -1;
        int carry = 0;
        int sum = 0;

        while (i >= 0 || j >=0) {
            sum = carry;
            if (i >= 0) sum += a.charAt(i) - '0';
            if (j >= 0) sum += b.charAt(j) - '0';

            sb.append(sum % 2);
            carry = sum / 2;

            i--; j--;
        }
        if (carry == 1)
            sb.append('1');
        return sb.reverse().toString();         
    }
}

File: 0067-add-binary/0067-add-binary.scala
object Solution {
    def addBinary(s1: String, s2: String): String = {
      println("Starting processing now...")
      val maxLen = Math.max(s1.length(),s2.length())

      var i = s1.length()-1
      var j = s2.length()-1
      var sum =0
      var carry = 0

      val sb= StringBuilder() 

      for(c <- 0 until maxLen){
          //println(s"c: $c")
          sum=sum + carry

          if(i >= 0)
            {
              println(s"Picked: ${s1.charAt(i)}")
              sum = sum + (s1.charAt(i) - '0')
              println(s"sum: $sum")
              i = i -1
            }
          
          if(j >= 0)
            {
              println(s"Picked: ${s2.charAt(j)}")
              sum = sum + (s2.charAt(j) - '0')
              println(s"sum: $sum")
              j =j-1
              }

          if(sum >= 2)
            carry = 1
          else 
            carry=0

          sb.append(sum % 2)
          sum = 0
          println(s"carry: $carry")
          println(s"Appended String: ${sb.toString()}")
          println("--------------------------------------")
      }
      if (carry == 1) sb.append(carry)
      sb.reverse.toString()
    }
}
